### ISSUE TITLE
PHP and HHVM version should not be inclusive

### ISSUE THREAD

**Issue by Soullivaneuh** [composer/composer#5442]
With the following composer.json:

```json
{
    "require": {
        "php": "^5.3 || ^7.0",
        "hhvm": "^3.12"
    }
}
```

When I run this command:

```
php /usr/local/bin/composer update -vvv
```

I get this output:

```
Reading ./composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/home/sullivan/tmp/composer): git branch --no-color --no-abbrev -v
Executing command (/home/sullivan/tmp/composer): git describe --exact-match --tags
Reading /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/linux-config/home/.composer/auth.json
Reading /home/sullivan/tmp/composer/vendor/composer/installed.json
Reading /home/sullivan/.composer/vendor/composer/installed.json
Loading plugin Pyrech\ComposerChangelogs\ChangelogsPlugin
Loading plugin SLLH\ComposerLint\LintPlugin
Loading plugin SLLH\ComposerVersionsCheck\VersionsCheckPlugin
Loading plugin Pyrech\ComposerNotifier\NotifierPlugin
Loading plugin Hirak\Prestissimo\Plugin
Downloading https://packagist.org/packages.json
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/packages.json into cache
    1/1:	http://packagist.org/p/provider-latest$17c9bd4e8a9cbaf3c99bb0e30b3a40b945f0515b70efc024937276a672b7ddae.json
    Finished: success: 1, skipped: 0, failure: 0, total: 1
Running 1.1.2 (2016-05-31 19:48:11) with PHP 5.6.4-4ubuntu6.4 on Linux / 3.19.0-49-generic
You are running composer with xdebug enabled. This has a major impact on runtime performance. See https://getcomposer.org/xdebug
Loading composer repositories with package information
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Resolving dependencies through SAT
Dependency resolution completed in 0.002 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - This package requires hhvm ^3.12 but you are running this with PHP and not HHVM.
```

And I expected this to happen:

Get the command working without any resolving issue.

I don't see why we can specify both php and hhvm version constraints if our project support both.

In this case, we can't add the `hhvm` part.

Furthermore, if I run the same command with `hhvm`:

```bash
hhvm /usr/local/bin/composer update -vvv
```

This works:

```bash
Reading ./composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file ./composer.json
Checked CA file /etc/ssl/certs/ca-certificates.crt: valid
Executing command (/home/sullivan/tmp/composer): git branch --no-color --no-abbrev -v
Executing command (/home/sullivan/tmp/composer): git describe --exact-match --tags
Reading /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/.composer/config.json
Loading config file /home/sullivan/.composer/auth.json
Loading config file /home/sullivan/.composer/composer.json
Loading config file /home/sullivan/linux-config/home/.composer/auth.json
Reading /home/sullivan/tmp/composer/vendor/composer/installed.json
Reading /home/sullivan/.composer/vendor/composer/installed.json
Loading plugin Pyrech\ComposerChangelogs\ChangelogsPlugin
Loading plugin SLLH\ComposerLint\LintPlugin
Loading plugin SLLH\ComposerVersionsCheck\VersionsCheckPlugin
Loading plugin Pyrech\ComposerNotifier\NotifierPlugin
Loading plugin Hirak\Prestissimo\Plugin
Downloading https://packagist.org/packages.json
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/packages.json into cache
    1/2:	http://packagist.org/p/provider-latest$814d76161463dc96c0734fbd83f5dbfef1d524286cd4ad27e08c41ee8a5d6c86.json
    2/2:	http://packagist.org/p/provider-2016-04$6766e3c7124bfb713689d907111aaf4478c2c6b3f01f5e03cabd17e37398b215.json
    Finished: success: 2, skipped: 0, failure: 0, total: 2
Running 1.1.2 (2016-05-31 19:48:11) with HHVM 3.14.1 on Linux / 3.19.0-49-generic
Loading composer repositories with package information
Writing /home/sullivan/.composer/cache/repo/https---packagist.org/packages.json into cache
Updating dependencies (including require-dev)
Resolving dependencies through SAT
Dependency resolution completed in 0.053 seconds
Analyzed 99 packages to resolve dependencies
Analyzed 99 rules to resolve dependencies
Nothing to install or update
Reading /home/sullivan/tmp/composer/vendor/composer/installed.json
Reading ./composer.lock
Generating autoload files
No changelogs summary
All packages are up to date.
```

This makes no sense, am I right?

**Comment by alcohol** [comment]
I think this is because while most php code will run on hhvm, code written specifically for hhvm will more than likely not run on php.

So in my opinion, the current situation makes sense.

**Comment by Seldaek** [comment]
It works with hhvm because hhvm provides a php version. Your project does not require hhvm if it works with plain php, so don't require hhvm, simple as that. The hhvm provided "php" version should match what functionality they support so that the php require is applied correctly.

**Comment by Soullivaneuh** [comment]
Thanks @alcohol and @Seldaek for the clarification. :+1:

**Comment by Soullivaneuh** [comment]
Another interrogation: Imagine HHVM as a bad implement of a targeted PHP feature in 3.9.0 and this is fixed on version 3.9.5.

How to specify the constraint on composer in this case?

Maybe on the `conflict` section?

**Comment by Seldaek** [comment]
Maybe conflict.. Or maybe it doesn't really matter because barely anyone
uses hhvm at this point.
